File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> s;
        for(int a : asteroids ){
            //if +ve :- right direction
            if(a > 0){
                s.push(a);

            }else{ // -ve :- left direction
                // when satck is non empty && stack have lesser value then remove it  
                while(!s.empty() && s.top() > 0 && s.top() < abs(a)){
                    s.pop();
                }
                // if  stack is empty or s have -ve value 
                if(s.empty() || s.top() < 0)s.push(a);
                // both have same value with opp direction :- both will be distroy
                else if(s.top() == abs(a))s.pop();

            }
        }
        vector<int>ans;
        while(!s.empty()){
            ans.push_back(s.top());
            s.pop();

        }
        reverse(ans.begin() , ans.end());
        return ans;
    }
};

File: 0735-asteroid-collision/0735-asteroid-collision.java
import java.util.Stack;

class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> asteroidStack = new Stack<>();

        for (int asteroid : asteroids) {
            if (asteroid > 0) {
                // Moving to the right, always push to stack
                asteroidStack.push(asteroid);
            } else {
                // Moving to the left, check for collisions
                while (!asteroidStack.isEmpty() && asteroidStack.peek() > 0 && asteroidStack.peek() < Math.abs(asteroid)) {
                    asteroidStack.pop();
                }

                if (asteroidStack.isEmpty() || asteroidStack.peek() < 0) {
                    asteroidStack.push(asteroid);
                } else if (asteroidStack.peek() == Math.abs(asteroid)) {
                    asteroidStack.pop();
                }
            }
        }

        // Convert stack back to array
        int[] result = new int[asteroidStack.size()];
        for (int i = asteroidStack.size() - 1; i >= 0; i--) {
            result[i] = asteroidStack.pop();
        }

        return result;
    }
}